File: a-star-cpp/src/random_map.cpp
#include <vector>
#include <array>
#include <cmath>
#include <cstdint>
#include "Point.h"  // 假设有Point类的定义
#include "random_map.h"
#include <cstdlib>
#include <ctime>

RandomMap::RandomMap(int size) : size(size) {
    map = std::vector<std::vector<std::array<uint8_t, 3>>>(size, std::vector<std::array<uint8_t, 3>>(size, { 255, 255, 255 }));
    GenerateObstacle();
}

// 返回地图尺寸
int RandomMap::GetSize() const {
    return size;
}

// 返回障碍物点的集合（常量引用，防止外部修改）
const std::vector<Point>& RandomMap::GetObstaclePoints() const {
    return obstacle_points;
}

// 返回地图数据（非常量引用，允许外部修改地图）
std::vector<std::vector<std::array<uint8_t, 3>>>& RandomMap::GetMap() {
    return map;
}

void RandomMap::GenerateObstacle() {
    obstacle_points.clear();

    // 添加水池障碍（中心圆形）
    Point fountain_center(240, 240);
    int fountain_radius = 40;
    AddCircleObstacle(fountain_center, fountain_radius);

    // 添加食品摊位障碍（正方形40x40）
    std::vector<Point> food_stalls = { {240, 120}, {160, 160}, {240, 340} };
    int food_width = 40;
    for (const auto& stall : food_stalls) {
        AddSquareObstacle(stall, food_width);
    }

    // 添加电梯障碍（正方形50x50）
    std::vector<Point> elevators = { {160, 400}, {320, 400} };
    int elevator_width = 50;
    for (const auto& elevator : elevators) {
        AddSquareObstacle(elevator, elevator_width);
    }

    // 添加购物商店障碍
    std::vector<Point> shopping_stores = { {0, 0}, {360, 0} };
    for (const auto& store : shopping_stores) {
        AddRectangleObstacle(store, 120, 480);
    }

    // 添加禁止区域障碍（顶部和底部水平区域）
    std::vector<Point> no_go_zone = { {120, 0}, {120, 440} };
    for (const auto& ngz : no_go_zone) {
        AddRectangleObstacle(ngz, 240, 40);
    }
}

void RandomMap::AddCircleObstacle(const Point& center, int radius) {
    for (int x = 0; x < size; ++x) {
        for (int y = 0; y < size; ++y) {
            if (std::pow(x - center.x, 2) + std::pow(y - center.y, 2) <= std::pow(radius, 2)) {
                map[x][y] = { 128, 128, 128 };  // 灰色障碍物
                obstacle_points.emplace_back(x, y);
            }
        }
    }
}

void RandomMap::AddSquareObstacle(const Point& center, int width) {
    int half_width = width / 2;
    for (int x = center.x - half_width; x < center.x + half_width; ++x) {
        for (int y = center.y - half_width; y < center.y + half_width; ++y) {
            if (IsValidCoordinate(x, y)) {
                map[x][y] = { 128, 128, 128 };  // 灰色障碍物
                obstacle_points.emplace_back(x, y);
            }
        }
    }
}

void RandomMap::AddRectangleObstacle(const Point& top_left, int width, int height) {
    for (int x = top_left.x; x < top_left.x + width; ++x) {
        for (int y = top_left.y; y < top_left.y + height; ++y) {
            if (IsValidCoordinate(x, y)) {
                map[x][y] = { 128, 128, 128 };  // 灰色障碍物
                obstacle_points.emplace_back(x, y);
            }
        }
    }
}

bool RandomMap::IsValidCoordinate(int x, int y) const {
    return x >= 0 && x < size && y >= 0 && y < size;
}

File: dijstra_py/random_map.py
# random_map.py

import numpy as np

import point

class RandomMap:
    def __init__(self, size=480):
        self.size = size  # 地图的尺寸
        self.obstacle_point = []
        self.map = np.ones((size, size, 3), dtype=np.uint8) * 255  # 初始化为白色
        self.GenerateObstacle()

    def GenerateObstacle(self):
        self.obstacle_point = []

        # 添加水池障碍（中心圆形）
        fountain_center = (240, 240)  # 中心坐标
        fountain_radius = 40
        for x in range(self.size):
            for y in range(self.size):
                if (x - fountain_center[0])**2 + (y - fountain_center[1])**2 <= fountain_radius**2:
                    self.map[x, y] = (128, 128, 128)  # 灰色障碍物
                    self.obstacle_point.append(point.Point(x, y))

        # 添加食品摊位障碍（正方形40x40）
        food_stalls = [(240, 120), (160, 160), (240, 340)]
        food_width = 40
        for stall in food_stalls:
            x_start, y_start = stall
            for x in range(x_start - 20, x_start + 20):
                for y in range(y_start - 20, y_start + 20):
                    self.map[x, y] = (128, 128, 128)  # 灰色障碍物
                    self.obstacle_point.append(point.Point(x, y))

        # 添加电梯障碍（正方形50x50）
        elevators = [(160, 400), (320, 400)]
        for elevator in elevators:
            x_start, y_start = elevator
            for x in range(x_start - 25, x_start + 25):
                for y in range(y_start - 25, y_start + 25):
                    self.map[x, y] = (128, 128, 128)  # 灰色障碍物
                    self.obstacle_point.append(point.Point(x, y))

        # 添加购物商店障碍 1
        shopping_stores = [(0, 0), (360, 0)]
        for store in shopping_stores:
            x_start, y_start = store
            for x in range(x_start, x_start + 120):
                for y in range(y_start, y_start + 480):
                    self.map[x, y] = (128, 128, 128)  # 灰色障碍物
                    self.obstacle_point.append(point.Point(x, y))

        # 添加禁止区域障碍（顶部和底部水平区域）
        no_go_zone = [(120, 0), (120, 440)]
        for ngz in no_go_zone:
            x_start, y_start = ngz
            for x in range(x_start, x_start + 240):  # 底部区域
                for y in range(y_start, y_start + 40):
                    self.map[x, y] = (128, 128, 128)  # 灰色障碍物
                    self.obstacle_point.append(point.Point(x, y))
